### PR TITLE
perf(ext/urlpattern): optimize URLPattern.exec

### PR DESCRIPTION
**Pull request by marcosc90**
This PR optimizes `URLPattern.exec` 

- Use component keys from constructor instead of calling it on every `.exec`. AFAIK keys should always be `protocol`,`username`,`password`,`hostname`,`port`,`pathname`,`search`,`hash`. Haven't looked much into it but I think it's safe to define these outside the constructor as well.
- Add a fast path for `/^$/u` (default regexp) and empty input
- Replaced `ArrayPrototypeMap` & `ObjectFromEntries` with a `for` loop.

**this PR**
```
cpu: 13th Gen Intel(R) Core(TM) i9-13900H
runtime: deno 1.36.1 (x86_64-unknown-linux-gnu)

benchmark      time (avg)        iter/s             (min … max)       p75       p99      p995
--------------------------------------------------------------- -----------------------------
exec 1          2.17 µs/iter     461,022.8     (2.14 µs … 2.27 µs)   2.18 µs   2.27 µs   2.27 µs
exec 2          4.13 µs/iter     242,173.4     (4.08 µs … 4.27 µs)   4.15 µs   4.27 µs   4.27 µs
exec 3          2.55 µs/iter     391,508.1     (2.53 µs … 2.68 µs)   2.56 µs   2.68 µs   2.68 µs
```

**main**
```
cpu: 13th Gen Intel(R) Core(TM) i9-13900H
runtime: deno 1.36.1 (x86_64-unknown-linux-gnu)

benchmark      time (avg)        iter/s             (min … max)       p75       p99      p995
--------------------------------------------------------------- -----------------------------
exec 1          2.45 µs/iter     408,092.4     (2.41 µs … 2.55 µs)   2.46 µs   2.55 µs   2.55 µs
exec 2          4.41 µs/iter     226,706.0   (3.49 µs … 399.56 µs)   4.39 µs   5.49 µs   6.07 µs
exec 3          2.99 µs/iter     334,833.4     (2.94 µs … 3.21 µs)   2.99 µs   3.21 µs   3.21 µs
```

### CODE REVIEW

**Comment by bartlomieju** [comment]
Asking @littledivy for review who has a WIP patch that optimizes `URLPattern` to use JS implementation.

**Review by littledivy — APPROVED** [review]
LGTM! I'm working on a different optimization: rewriting JS version speacialized for .text & .exec `op_urlpattern_process_match_input`

File: ext/url/01_urlpattern.js [hunk 1]
```diff
@@ -64,19 +64,21 @@ class URLPattern {
 
     const components = ops.op_urlpattern_parse(input, baseURL);
 
-    const keys = ObjectKeys(components);
-    for (let i = 0; i < keys.length; ++i) {
-      const key = keys[i];
+    this.#keys = ObjectKeys(components);
```
Comment (lucacasonato) [inline ext/url/01_urlpattern.js:67]: The keys are a fixed array I think  (`protocol`, `hostname`, etc)

I wonder if we even need to do any `ObjectKeys` and store in a field, or if we can just use a const global array.

Comment (marcosc90) [reply ext/url/01_urlpattern.js]: That was my initial thought. I pushed the change to avoid `ObjectKeys` in the constructor

**Pull request merged**